File: heuristics.py
print ':- module(heuristics,'
print '        [ first_move/2,          % The heuristic for first move'
print '          value/3                % The heuristic for board state'
print '        ]).'
print ''
print ':- use_module(board,'
print '        [ me/1,'
print '          opponent/2,'
print '          empty_board/1,'
print '          moves/2'
print '        ]).'
print ''
print '%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%'
print '%%%'
print '%%%     Heuristics'
print '%%%'
print '%%%'
print '%%%     The h_func calculates the heuristics function for a'
print '%%%     given position of the board.'
print '%%%'
print '%%%         h_func(+Board, +Player, -Weight)'
print '%%%'
print '%%%         where Board is the current board state'
print '%%%               Player is the piece of current player (x/o)'
print '%%%               Weight is the heuristic value'
print ''
print 'not_opponent(A,B,C,D,x) :-'
print '  (A = x; A = 0),'
print '  (B = x; B = 0),'
print '  (C = x; C = 0),'
print '  (D = x; D = 0).'
print ''
print 'not_opponent(A,B,C,D,o) :-'
print '  (A = o; A = 0),'
print '  (B = o; B = 0),'
print '  (C = o; C = 0),'
print '  (D = o; D = 0).'
print ''
print 'found(0,0) :- !.'
print 'found(_,1) :- !.'
print ''
print 'w(0,0) :- !.      % no pieces, then 0'
print 'w(1,100) :- !.    % one piece, then 100'
print 'w(2,1000) :- !.   % two pieces, then 1000'
print 'w(3,10000) :- !.  % three pieces, then 10000'
print 'w(4,100000) :- !. % four pieces, then 100000'
print ''
print 'weight(A,B,C,D,Player,W) :-'
print '  ('
print '    not_opponent(A,B,C,D,Player) ->'
print '    ('
print '      found(A,W1),'
print '      found(B,W2),'
print '      found(C,W3),'
print '      found(D,W4),'
print '      S is W1+W2+W3+W4,'
print '      w(S,W)'
print '    ) ;'
print '    W = 0'
print '  ), !.'
print ''

class Board:
	def __init__(self, pieces = []):
		self.pieces=pieces

	def __str__(self):
		return " /\n\n".join([" /\n".join(["  " + "/".join(line) for line in plane]) for plane in self.board()])

	def board(self):
		N=['W','Z','Y','X']
		pieces=[]
		for k in range(0,4):
			plane=[]
			for j in range(0,4):
				line=[]
				for i in range(0,4):
					line.append('%s%d%d' % (N[k],j,i))
				plane.append(line)
			pieces.append(plane)
		pieces.reverse()
		return pieces

def wins():
	w=[]
	w.append([(3,0,0), \
			  (2,1,1), \
			  (1,2,2), \
			  (0,3,3)])
	w.append([(3,0,3), \
			  (2,1,2), \
			  (1,2,1), \
			  (0,3,0)])
	w.append([(3,3,3), \
			  (2,2,2), \
			  (1,1,1), \
			  (0,0,0)])
	w.append([(3,3,0), \
			  (2,2,1), \
			  (1,1,2), \
			  (0,0,3)])
	for j in range(0,4):
		w.append([(3,0,j), \
				  (2,1,j), \
				  (1,2,j), \
				  (0,3,j)])
		w.append([(3,3,j), \
				  (2,2,j), \
				  (1,1,j), \
				  (0,0,j)])
		w.append([(3,j,0), \
				  (2,j,1), \
				  (1,j,2), \
				  (0,j,3)])
		w.append([(3,j,3), \
				  (2,j,2), \
				  (1,j,1), \
				  (0,j,0)])
		w.append([(j,3,0), \
				  (j,2,1), \
				  (j,1,2), \
				  (j,0,3)])
		w.append([(j,3,3), \
				  (j,2,2), \
				  (j,1,1), \
				  (j,0,0)])
		for i in range(0,4):
			w.append([(0,i,j), \
					  (1,i,j), \
					  (2,i,j), \
					  (3,i,j)])
			w.append([(i,0,j), \
					  (i,1,j), \
					  (i,2,j), \
					  (i,3,j)])
			w.append([(i,j,0), \
					  (i,j,1), \
					  (i,j,2), \
					  (i,j,3)])
	return w

def intercepts(wins, piece):
	r=[]
	for board in wins:
		if piece in set(board):
			r.append(board)
	return r

def merge((z,y,x), boards):
	i='A'
	m=dict()
	m[(z,y,x)]='_X'
	for board in boards:
		j=0
		for piece in board:
			if piece not in set(m.keys()):
				m[piece]='%s%d' % (i,j)
				j+=1
		i=chr(ord(i) + 1)
	return m

def single_merge(board):
	i='A'
	m=dict()
	for piece in board:
		m[piece]='%s' % i
		i=chr(ord(i) + 1)
	return m

def print_h(board, i):
	line=[]
	for (z,y,x) in board.keys():
		N=['W','Z','Y','X']
		line.append('%s%d%d' % (N[z],y,x))

	print '  weight(' + ','.join(line) + ',Player,S%d),' % i

def print_crosses((z,y,x), weight):
	print 'crosses([%d, %d, %d], %d).' % (z,y,x,weight)

def heuristics():
	w=wins()

	print 'h_func('
	print str(Board()) + ','
	print '  Player, W) :- '
	print ''

	i = 1
	for board in w:
		M=single_merge(board)
		print_h(M, i)
		i += 1

	l=[]
	for j in range(0,i-1):
		l.append('S%d' % (j+1))

	print '  W is ' + '+'.join(l) + ', !.'
	print ''

	for z in range(0,4):
		for y in range(0,4):
			for x in range(0,4):
				I=intercepts(w, (z,y,x))
				M=merge((z,y,x),I)
				print_crosses((z,y,x),len(I))
	print ''

heuristics()

print 'first_move(BestMove, BestValue) :-'
print '  board:empty_board(E),'
print '  board:moves(E, Moves),'
print '  first_move(Moves, nil/0, BestMove/BestValue), !.'
print ''
print 'first_move([], Move/Val, Move/Val) :- !.'
print ''
print 'first_move([Move|TailMoves], SelMove/Val, BestMove/BestValue) :-'
print '  random(1,1000, R),'
print '  ((crosses(Move, X), X = 7) ->'
print '    (Score is 1000 + R) ; (Score is 100 + R)),'
print '  first_move(Move/Score, SelMove/Val, SelMove1/Val1),'
print '  first_move(TailMoves, SelMove1/Val1, BestMove/BestValue), !.'
print ''
print 'first_move(_/Val1, Move2/Val2, Move2/Val2) :-'
print '  Val2 > Val1, !.'
print ''
print 'first_move(Move1/Val1, _, Move1/Val1) :- !.'
print ''
print 'value(Board, Player, Val) :-'
print '  board:me(Me),'
print '  board:opponent(Me, Opponent),'
print '  h_func(Board, Me, W1),'
print '  h_func(Board, Opponent, W2),'
print '  scoreof(W1, W2, Player, Val), !.'
print ''
print 'scoreof(W1, W2, Player, Val) :-'
print '  board:me(Player), Val is W1-W2, !.'
print ''
print 'scoreof(W1, W2, _, Val) :-'
print '  Val is W2-W1, !.'
print ''

File: heuristics.pl
:- module(heuristics,
        [ first_move/2,          % The heuristic for first move
          value/3,               % The heuristic for board state
          win/2                  % Whether some player wins
        ]).

:- use_module(board,
        [ me/1,
          opponent/2,
          empty_board/1,
          moves/2
        ]).

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%
%%%     Heuristics
%%%
%%%
%%%     The h_func calculates the heuristics function for a
%%%     given position of the board.
%%%
%%%         h_func(+Board, +Player, -Weight)
%%%
%%%         where Board is the current board state
%%%               Player is the piece of current player (x/o)
%%%               Weight is the heuristic value

weight(0, 0).      % no pieces, then 0
weight(1, 100).    % one piece, then 100
weight(2, 1000).   % two pieces, then 1000
weight(3, 10000).  % three pieces, then 10000
weight(4, 900000). % four pieces, then 900000

score(A, B, C, D, Player, W) :-
  in_row([A, B, C, D], Player, 0, Sum),
  weight(Sum, W)
  ;
  W = 0.

in_row([], _, Sum, Sum).

in_row([H|T], Player, Aux, Sum) :-
  H = Player -> Aux1 is Aux+1 ;
  H = 0 -> Aux1 is Aux,
  find(T, Player, Aux1, Sum).

wins(
  X00/X01/X02/X03 / X10/X11/X12/X13 / X20/X21/X22/X23 / X30/X31/X32/X33 /
  Y00/Y01/Y02/Y03 / Y10/Y11/Y12/Y13 / Y20/Y21/Y22/Y23 / Y30/Y31/Y32/Y33 /
  Z00/Z01/Z02/Z03 / Z10/Z11/Z12/Z13 / Z20/Z21/Z22/Z23 / Z30/Z31/Z32/Z33 /
  W00/W01/W02/W03 / W10/W11/W12/W13 / W20/W21/W22/W23 / W30/W31/W32/W33,

  [ % all wins for a 4x4x4x4 board
    (Z22,W33,X00,Y11), (W30,Z21,Y12,X03), (X33,Y22,Z11,W00), (Y21,X30,W03,Z12),
    (W30,X00,Y10,Z20), (X30,Z10,W00,Y20), (X00,Z02,Y01,W03), (Y02,Z01,W00,X03),
    (W30,W12,W21,W03), (W11,W33,W00,W22), (Z00,X00,W00,Y00), (W30,W20,W10,W00),
    (W00,W01,W02,W03), (Z10,W10,X10,Y10), (Z00,Z10,Z20,Z30), (Z00,Z03,Z02,Z01),
    (X20,W20,Z20,Y20), (Y10,Y00,Y30,Y20), (Y02,Y03,Y00,Y01), (W30,X30,Z30,Y30),
    (X20,X30,X00,X10), (X02,X01,X00,X03), (X01,Y11,W31,Z21), (Y21,X31,W01,Z11),
    (Z12,X10,W13,Y11), (Z11,W10,Y12,X13), (Z03,Z21,Z30,Z12), (Z22,Z00,Z11,Z33),
    (X01,W01,Y01,Z01), (W11,W21,W01,W31), (W12,W11,W10,W13), (W11,Z11,Y11,X11),
    (Z21,Z11,Z31,Z01), (Z13,Z10,Z11,Z12), (Y21,Z21,W21,X21), (Y21,Y11,Y31,Y01),
    (Y11,Y10,Y13,Y12), (X31,Z31,Y31,W31), (X01,X31,X11,X21), (X10,X11,X12,X13),
    (X02,Z22,W32,Y12), (Y22,Z12,W02,X32), (X20,Y21,W23,Z22), (X23,W20,Z21,Y22),
    (Y21,Y03,Y12,Y30), (Y22,Y33,Y00,Y11), (X02,Z02,Y02,W02), (W12,W32,W02,W22),
    (W23,W20,W21,W22), (W12,X12,Y12,Z12), (Z22,Z02,Z32,Z12), (Z22,Z21,Z20,Z23),
    (Z22,Y22,X22,W22), (Y22,Y02,Y32,Y12), (Y21,Y22,Y23,Y20), (W32,Y32,Z32,X32),
    (X02,X22,X12,X32), (X20,X23,X22,X21), (W33,Z23,Y13,X03), (X33,Z13,Y23,W03),
    (X30,W33,Y31,Z32), (X33,Z31,Y32,W30), (X30,X03,X12,X21), (X33,X22,X11,X00),
    (Z03,Y03,W03,X03), (W23,W33,W03,W13), (W30,W33,W32,W31), (Z13,Y13,X13,W13),
    (Z13,Z03,Z33,Z23), (Z30,Z31,Z32,Z33), (W23,X23,Y23,Z23), (Y23,Y33,Y03,Y13),
    (Y33,Y32,Y31,Y30), (X33,W33,Y33,Z33), (X33,X23,X13,X03), (X33,X30,X31,X32)
  ]).

h_func(Board, Player, W) :-
  wins(Board, Stack),
  h_sum(Stack, Player, 0, W).

h_sum([], _, W, W).

h_sum([(A,B,C,D)|Tail], Player, S, W) :-
  score(A, B, C, D, Player, Score),
  S1 is S + Score,
  h_sum(Tail, Player, S1, W).

crosses([Z,Y,X],C) :-
  (
    [Z,Y,X] = [0,0,0]; [Z,Y,X] = [0,0,3]; [Z,Y,X] = [0,3,0]; [Z,Y,X] = [0,3,3];
    [Z,Y,X] = [1,1,1]; [Z,Y,X] = [1,1,2]; [Z,Y,X] = [1,2,1]; [Z,Y,X] = [1,2,2];
    [Z,Y,X] = [2,1,1]; [Z,Y,X] = [2,1,2]; [Z,Y,X] = [2,2,1]; [Z,Y,X] = [2,2,2];
    [Z,Y,X] = [3,0,0]; [Z,Y,X] = [3,0,3]; [Z,Y,X] = [3,3,0]; [Z,Y,X] = [3,3,3]
  )
  -> C = 7 ; C = 4.

win(Board, Player) :-
  wins(Board, Stack),
  find_win(Stack, Player) ; fail.

find_win([(A,B,C,D)|Tail], Player) :-
  in_row([A, B, C, D], Player, 0, Sum),
  Sum = 4 -> true ; find_win(Tail, Player).

first_move(BestMove, BestValue) :-
  board:empty_board(E),
  board:moves(E, Moves),
  first_move(Moves, nil/0, BestMove/BestValue), !.

first_move([], Move/Val, Move/Val) :- !.

first_move([Move|TailMoves], SelMove/Val, BestMove/BestValue) :-
  random(1,1000, R),
  ((crosses(Move, X), X = 7) ->
    (Score is 1000 + R) ; (Score is 100 + R)),
  first_move(Move/Score, SelMove/Val, SelMove1/Val1),
  first_move(TailMoves, SelMove1/Val1, BestMove/BestValue), !.

first_move(_/Val1, Move2/Val2, Move2/Val2) :-
  Val2 > Val1, !.

first_move(Move1/Val1, _, Move1/Val1).

value(Board, Player, Val) :-
  board:me(Me),
  board:opponent(Me, Opponent),
  h_func(Board, Me, W1),
  h_func(Board, Opponent, W2),
  scoreof(W1, W2, Player, Val).

scoreof(W1, W2, Player, Val) :-
  board:me(Player), Val is W1-W2, !.

scoreof(W1, W2, _, Val) :-
  Val is W2-W1.

